mm2src/coins/utxo/rpc_clients/electrum_rpc/connection.rs

```diff
@@ -0,0 +1,872 @@
+    }
+
+    /// Process an incoming JSONRPC response from the electrum server.
+    async fn process_electrum_response(
+        &self,
+        raw_json: Json,
+        event_handlers: &Vec<Box<SharableRpcTransportEventHandler>>,
+    ) {
+        // Inform the event handlers.
+        if let Ok(ref data) = json::to_vec(&raw_json) {
+            event_handlers.on_incoming_response(data);
+        }
+
+        // detect if we got standard JSONRPC response or subscription response as JSONRPC request
+        #[derive(Deserialize)]
+        #[serde(untagged)]
+        enum ElectrumRpcResponseEnum {
+            /// The subscription response as JSONRPC request.
+            ///
+            /// NOTE Because JsonRpcResponse uses default values for each of its field,
+            /// this variant has to stay at top in this enumeration to be properly deserialized
+            /// from serde.
+            SubscriptionNotification(JsonRpcRequest),
+            /// The standard JSONRPC single response.
+            SingleResponse(JsonRpcResponse),
+            /// The batch of standard JSONRPC responses.
+            BatchResponses(JsonRpcBatchResponse),
+        }
+
+        let response: ElectrumRpcResponseEnum = match json::from_value(raw_json) {
+            Ok(res) => res,
+            Err(e) => {
+                error!("{}", e);
+                return;
+            },
+        };
+
+        let response = match response {
+            ElectrumRpcResponseEnum::SingleResponse(single) => JsonRpcResponseEnum::Single(single),
+            ElectrumRpcResponseEnum::BatchResponses(batch) => JsonRpcResponseEnum::Batch(batch),
+            ElectrumRpcResponseEnum::SubscriptionNotification(req) => {
+                // NOTE: Sending a script hash notification is handled in it's own event handler.
+
+                // FIXME: What is this used for? Note that the id is the method name in this case (two similar id will collide),
+                // but this isn't a response to any request anyways, this is a notification, and we forwarded using the
+                // scripthash_notification_sender above already.
+                JsonRpcResponseEnum::Single(JsonRpcResponse {
+                    id: req.method.clone(),
+                    jsonrpc: "2.0".into(),
+                    result: req.params[0].clone(),
+                    error: Json::Null,
+                })
+                //return; // BLOCKCHAIN_HEADERS_SUB_ID wasn't handled, just returned.
+                //also, you might want to check the notification type to print an error message if it's not expected
+            },
+        };
+
+        // the corresponding sender may not exist, receiver may be dropped
+        // these situations are not considered as errors so we just silently skip them
+        let pending = self.responses.lock().await.remove(&response.rpc_id());
+        if let Some(tx) = pending {
+            tx.send(response).ok();
+        }
```

Comment:
I am not sure why it was done this way and why it causes no problems, I guess because it's subscriptions but still why no error is returned here
https://github.com/KomodoPlatform/komodo-defi-framework/blob/065213b87ae1a924e3424664844d356ec754db15/mm2src/coins/utxo/utxo_balance_events.rs#L63-L67
When we we will be waiting on the response on the channel related to the request_id
https://github.com/KomodoPlatform/komodo-defi-framework/blob/df6ab98b3b2ee84918bfc0bf1bc03d1e1f219985/mm2src/coins/utxo/rpc_clients.rs#L3064-L3069
I guess it's related to how we handle subscriptions here https://github.com/KomodoPlatform/komodo-defi-framework/blob/065213b87ae1a924e3424664844d356ec754db15/mm2src/coins/utxo/utxo_balance_events.rs#L100-L119 and that we will get the Triggered variant anyway. I think we should fix this but I am not sure why it was done this way and if it will cause problems if fixed. Only reason I can think of is that it was done this way to not allow concurrent subscriptions since one is enough, but we send the subscriptions either way but not wait on the response/s.

BTW, all code snippets are from dev branch, sorry about that as I was tracing this in dev. What is your opinion on this @mariocynicys @onur-ozkan , should we fix this? I would advise to fix this @mariocynicys but test it before and after fixing and try to notice any differences in logs etc..

Reply:
`blockchain_scripthash_subscribe_using` doesn't use the method name as `id`:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/260192eb8373a35f226f2280a21eee17ebf33f30/mm2src/coins/utxo/rpc_clients/electrum_rpc/client.rs#L804-L809

The fixme linked above is regarding to the trigger and not to the sub request. I was thinking it's just a mistake by storing the result in our response map, but wanna make sure.

Reply:
> I am not sure why it was done this way and why it causes no problems, I guess because it's subscriptions but still why no error is returned here

Well, this part will never finish now until the addresses are subscribed:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/260192eb8373a35f226f2280a21eee17ebf33f30/mm2src/coins/utxo/rpc_clients/electrum_rpc/connection_managers/mod.rs#L171-L202

So if we couldn't subscribe we will keep trying endlessly. This can block utxo event streaming handler if we are going through some network issues, which will stack the re-subscribe messages until we could successfully start subscribing again.